include/deal.II/base/symmetric_tensor.h

```diff
@@ -3247,7 +3248,8 @@ deviator_tensor()
   // fill the elements treating the diagonal
   for (unsigned int i = 0; i < dim; ++i)
     for (unsigned int j = 0; j < dim; ++j)
-      tmp.data[i][j] = Number((i == j ? 1 : 0) - 1. / dim);
+      tmp.data[i][j] =
+        internal::NumberType<Number>::value((i == j ? 1 : 0) - 1. / dim);
```

Comment:
wouldn't it make more sense to write 1. and 0. (here and in other places)?

Reply:
I don't think it matters -- the `NumberType<T>::value` functions generally take a `T` (or related scalar) as argument, so it is automatically converted if you write `1` instead of `1.0`. Here, it is converted to `double` because you subtract a `double` from an `int`.

But I'll let @jppelteret comment on this as well.

Reply:
I think that what @tjhei is suggesting is that we write the first term of `1-1.0/dim` as a double so that we don't need to do an unnecessary implicit `int`->`double` conversion there. I don't know enough about compilers to understand if it would do this by itself and without a performance loss. I'd be happy to follow though with this if there's something to be gained by doing so.

Reply:
In the other places, where we do (or want to do) `internal::NumberType<Number>::value(1)`, then I'd say that @bangerth's description of the conversion is correct. But we never use `SymmetricTensor<rank,dim, int>`, right? So we could again just circumvent the need for a conversion in the most common case, that where the `Number` is a `double`, by initialising with a double in the first place.

Reply:
For constants, the conversion will actually be done at compile time. I don't think anyone suggested that the change would make things faster or more correct, it's just a matter of stating explicitly what actually happens implicitly. I think it's not a big deal here.

The only case where I could come up with an issue is if you had a type `FP` and wanted to create a `SymmetricTensor<2,dim,FP>` and `FP` has two constructors of which one takes a `double` and the other `int`, and the latter does something completely different than the former and not what we actually want. That's a bit hard to imagine, and would surely be poor design.

I'm tempted to say that we should just merge, and that if anyone wanted to be more explicit, that we address this in a follow-up. It's for sure unrelated to what this patch does in particular.